### PR TITLE
feat: add Gitpod support

### LINKED ISSUE #2347: ✨ Add Github Gitpod configuration
### What are you trying to do?

I'm testing out using Gitpod to write contributions for the Dagger repo, but the default Gitpod environment is general purpose and so isn't tailored for this specific repo. It would be useful to have a fixed Gitpod configuration for this repo for repeatable dev environments on any branch.

### Why is this important to you?

I can easily work on this repo without having to make sure I have my dev environment set up correctly. Additionally every branch I work on will always have a version of Dagger compiled from that branch's head.

### How are you currently working around this?

I have to create new config files for each new branch and make sure I don't commit them if making a pull request.

### PR DESCRIPTION
**Pull request by berryp**
Signed-off-by: Berry Phillips <berryphillips@gmail.com>

Closes #2347

This environment includes all Gitpod supported languages and tools, as well as some additional terminal tools that are popular.

I went back and forth on using the full base image vs. the slimmer Go image. The full image contains all supported languages and tools - including Docker ([see here](https://github.com/gitpod-io/workspace-images/tree/main/chunks)) - whilst the Go image contains just Go. Using a custom Docker file to install Docker and other desirable tools just adds to load times and complexity. I also think there is merit in having other languages and tools available to devs in the environment. So, I opted for the full image.

[Open in Gitpod](https://gitpod.io/#https://github.com/berryp/dagger/tree/2347-add-gitpod-support)

### CODE REVIEW

File: ci.cue
Comment (aluzzardi) [inline ci.cue:111]: This is simply to verify `gitpod` works alright, correct? (e.g. we're not using the resulting image in any way)

File: .gitpod.yml
Comment (aluzzardi) [inline .gitpod.yml:7]: Instead of this, personally I'm using these two from the marketplace:

https://marketplace.visualstudio.com/items?itemName=nickgo.cuelang
https://marketplace.visualstudio.com/items?itemName=jallen7usa.vscode-cue-fmt

**Review by aluzzardi — COMMENTED** [review]
Thank you!

I believe there's a linting error in CUE, you can fix this by running `make cuefmt` (or using the VSCode extension I mentioned :)

File: .gitpod.yml
Comment (berryp) [reply .gitpod.yml:7]: Gitpod can't use extensions from the Microsoft marketplace, due to licensing restrictions https://www.gitpod.io/docs/ides-and-editors/vscode-extensions#where-do-i-find-extensions. The two extensions you mentioned are also what I use locally but they are not listed on Open VSX yet. I've added issues to each repo to do so, but that could take some time.

File: ci.cue
Comment (berryp) [reply ci.cue:111]: Correct. I'll add a comment in the code to clarify.

File: .gitpod.yml
Comment (berryp) [reply .gitpod.yml:7]: I'll comment the code to make that clear.

File: ci.cue
Comment (berryp) [reply ci.cue]: I renamed the plan to "Test" for clarity.

File: .github/workflows/test-gitpod.yml
Comment (TomChv) [inline .github/workflows/test-gitpod.yml:13]: You should also rerun gitpod on changes on dagger engine/universe/core

File: .github/workflows/test-gitpod.yml
Comment (TomChv) [inline .github/workflows/test-gitpod.yml:39]: Remove those one for now, we must use a hack to avoid overriding cache from other PR.
See #2120 for details

File: .gitpod.yml
Comment (TomChv) [inline .gitpod.yml:27]: Is there a way to config a sign-off? Dagger contribution guidelines ask that all contributors should signs their commits.

**Review by TomChv — COMMENTED** [review]
I tested it and that's really amazing!

I left comments about ci, we can merge it after that

File: .gitpod.yml
Comment (berryp) [reply .gitpod.yml]: I changed the alias to `commit -s` which does the job.

**Comment by berryp** [comment]
I optimised the image somewhat as build times were causing delays when trying to open a Gitpod:

 - The `full` image size is around 7GB which takes a while to download
 - Homebrew takes forever to download all of its formulae. I couldn't find any option to prevent it downloading them
 
Changes I made:

 - Use the Go base image. It's still quite big - around 2GB - but a lot quicker to pull down
 - Install packages with apt, rather than Homebrew
 - I removed some of the packages as they are were not made available until Ubuntu 20.10 and I didn't want to overcomplicate this initial PR with a bunch ton of install commands

Overall, time to editor should be a fair bit quicker.

**Comment by berryp** [comment]
> Can you rebase your PR to clean up git history? There are changes from other's PR in

I rebased off origin, not upstream 🤦🏻‍♂️

**Review by TomChv — APPROVED** [review]
LGTM! We should merge that one after another approval!

**Review by shykes — APPROVED** [review]

**Comment by shykes** [comment]
Hi @berryp , thanks for the contribution! There was a small conflict, so I took the liberty of resolving it, rebasing and force-pushing. As soon as tests finish running, I'll merge.

**Pull request merged**
